src/bssl.rs

```diff
@@ -12,12 +12,62 @@
 // OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 // CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
-use {c, error};
+pub use self::ffi_error::ErrorCode;
 
-pub fn map_result(bssl_result: c::int) -> Result<(), error::Unspecified> {
-    match bssl_result {
-        1 => Ok(()),
-        _ => Err(error::Unspecified),
+/// Many foreign functions return a `c::int` indicating whether an error occurred. This module
+/// wraps that error code in a `#[repr(transparent)]` newtype which can be converted into a
+/// `Result`.
+mod ffi_error {
+    use core::marker::PhantomData;
+    use {c, error};
+
+    /// The meaning of the error code returned by a foreign function.
+    pub trait Semantics {
+        /// Returns true if a given error code indicates success.
+        fn is_success(code: c::int) -> bool;
+    }
+
+    /// The semantics used by most BoringSSL functions:
+    ///
+    /// * `1` indicates success.
+    /// * `0` indicates failure.
+    #[derive(Debug, Clone, Copy)]
+    pub enum BoringSsl {}
+
+    impl Semantics for BoringSsl {
+        fn is_success(code: c::int) -> bool {
```

Comment:
I would say, rather than defining an `is_success() -> bool` function, just inline this logic directly into the `From` implementation that does the conversion to `Result<(), error::Unspecified>`.